### ISSUE TITLE
Feature: unique_usertype_traits CHECK extension point

### ISSUE THREAD

**Issue by Rochet2** [ThePhD/sol2#1010]
1. Produce any relevant imaginary code that illustrates what you would like or desired behavior.

https://godbolt.org/z/b4qs6e

2. Include a description and title of what you would like.

I am currently trying to create safety around raw pointers.
I cannot alter the entire API and underlying systems of someone else's program to use shared_ptr, so instead I build on top of existing code to try create safety.
The approach I have is similar to shared_ptr, except that I store a shared_ptr into an object and use a wrapper containing weak_ptr from lua to be able to check if it still exists.
Sol does not seem to provide the means to do this checking if unique_usertype_traits is used, which seems to be required to allow the use of a wrapper properly with inheritance.

Alternative ways to do the same safety for raw pointers are appreciated as well :)

Additionally, as I am trying to avoid the use of unsafe pointers / add safety around them I tried seeing if there was a way to block pushing of raw pointers into lua, but I could not find such an option from sol.

3. Annotate and describe the behavior through comments, asserts or just a small write up.

I store a shared_ptr to the classes that I need safety for:
```cpp
class RawPtrCapable
{
public:
    std::shared_ptr<bool> exists = std::make_shared<bool>();
};

class SomeClass1 : public RawPtrCapable
{
public:
   virtual std::string foo() { return "bar1"; }
};

class SomeClass2 : public SomeClass1
{
public:
   virtual std::string foo() override { return "bar2"; }
};
```

I have the classes properly exposed with bases for inheritance.
```cpp
        {
            auto mt = lua.new_usertype<SomeClass1>("SomeClass1");
            mt["foo"] = static_cast<std::string(SomeClass1::*)()>(&SomeClass1::foo);
        }
        {
            auto mt = lua.new_usertype<SomeClass2>("SomeClass2", sol::base_classes, sol::bases<SomeClass1>());
            mt["foo"] = static_cast<std::string(SomeClass2::*)()>(&SomeClass2::foo);
        }
```

When pushing a raw pointer of that class to lua I wrap it inside a custom class, which is exposed as a wrapper similar to shared_ptr so I can use the inheritance capabilities of those exposed classes. The wrapper contains a weak_ptr, which can be used to check if the object was deleted or not.
```cpp
template<typename T>
class raw_ptr
{
public:
    raw_ptr(T* t) : rawPtr(t), weak_ptr(t->exists) {
    }

    bool is_null() const {
        std::cout << "IS_NULL REAL" << std::endl;
        return weak_ptr.expired();
    }

    T* get() const {
        std::cout << "GET REAL" << std::endl;
        return rawPtr;
    }

private:
    std::weak_ptr<bool> weak_ptr;
    T* rawPtr;
};

namespace sol {
    template <typename T>
    struct unique_usertype_traits<raw_ptr<T>> {
        typedef T type;
        typedef raw_ptr<T> actual_type;
        // template <typename X>
        // using rebind_base = raw_ptr<X>;

        static const bool value = true;

        static bool is_null(const actual_type& p) {
            std::cout << "IS_NULL" << std::endl;
            return p.is_null();
        }

        static type* get(const actual_type& p) {
            std::cout << "GET" << std::endl;
            return p.get();
        }

        // PROPOSAL
        static bool check(const actual_type& p) {
            std::cout << "CHECK" << std::endl;
            return p.is_null();
        }
    };
}
```

In the lua code I try to use the wrapped pointer as a parameter to a function that accepts `SomeClass1`...
```cpp
void test(SomeClass1* rawPtr) {
    std::cout << rawPtr->foo() << std::endl;
}

        SomeClass2* rawPtr = new SomeClass2();

        lua["rawptr"] = createSolRawRef(rawPtr);
        // delete rawPtr;

        lua["test"] = test;
        std::cout << "STARTING FUNCTION CALL" << std::endl;
        lua.safe_script("test(rawptr)");
        std::cout << "ENDING FUNCTION CALL" << std::endl;
```

When the code is run I get the following output.
```
IS_NULL
IS_NULL REAL
GET
GET REAL
STARTING FUNCTION CALL
bar2
ENDING FUNCTION CALL
```

From this output we can see that during the function call there was no checking of any kind done to the object - which is actually not needed for unique_ptr or shared_ptr as they can be checked already when pushing into lua.
It seems that it is not possible to tap into the CHECK of `unique_usertype_traits` objects, which is what I need in order to test if the weak_ptr is still valid and throw error if its not when the actual value is being fetched back into C++.

I am unsure if there are better ways to do this kind of thing, but this is somewhat the only way that comes to mind.
So as such, I would propose additional extension points where it is possible to tap into the CHECK call of a class using unique_usertype_traits in order to expand the checking capabilities.

**Comment by Rochet2** [comment]
The explanation above is probably not the clearest. So here is a bit more clear explanation.
What I try to do is make the following code safe
```cpp
SomeClass2* rawPtr = new SomeClass2();
lua["rawptr"] = rawPtr;
delete rawPtr; // The pointer can be deleted when lua is not executed.
lua.safe_script("test(rawptr)"); // we want to raise an error here, not crash or have undefined behavior.
```

My approach is to do it by using a `std::weak_ptr` style wrapper that uses `unique_usertype_traits` similar to `std::shared_ptr`.
The wrapper's check for the existence of the underlying value should be done every time the value is used by a C++ function, so there is a need for a CHECK extension point for `unique_usertype_traits`.

The extension point would allow building the following
```cpp
SomeClass2* rawPtr = new SomeClass2();
lua["rawptr"] = createSolRawRef(rawPtr);
delete rawPtr; // The pointer can be deleted when lua is not executed.
lua.safe_script("test(rawptr)"); // the wrapper will check if a shared_ptr still exists and will throw.
```

The approach is sketched in the godbolt link code.

I first thought the extension point could maybe be used to implement `std::weak_ptr` in general, but as explained here: https://github.com/ThePhD/sol2/issues/129#issuecomment-228642961
the lock() call of a weak_ptr cannot be easily implemented.

**Comment by ThePhD** [comment]
This is going to be part of a larger push to allow users to heavily customize how usertypes are worked with and interpreted through sol2. I'm not sure I'll be able to work it in exactly like this, but it should still work.

One of the problem is my use of these functions internally in sol2, which fix their template instantiation and refuse to allow you to do things like e.g. override the use of `std::size_t` in the internal framework properly. It's going to take some finesse on my end to make this structure work...

**Comment by ThePhD** [comment]
I am going to break this off into a smaller feature rather than a complete overhaul to ship something usable and future-proof, today, so that #1074 can be fixed and _most_ concerns here can be addressed.

**Comment by ThePhD** [comment]
An implementation that can get this done is available in #1074. I'm going to put the documentation part of this into a mega issue for v4/docs.
